File: Python/detect_capital.py
# https://leetcode.com/problems/detect-capital/description/
from utils.examination import Examiner, Testcase


class Solution:
    def detectCapitalUse(self, word: str) -> bool:
        if 2 > len(word):
            return True
        if word[0].islower() and not word[1].islower():
            return False
        is_lower: bool = word[1].islower()
        for i in range(2, len(word)):
            if not word[i].islower() == is_lower:
                return False
        return True

def examine() -> None:
    Examiner(Solution(), 'detectCapitalUse').run((
        Testcase(('',), True),
        Testcase(('g',), True),
        Testcase(('gg',), True),
        Testcase(('G',), True),
        Testcase(('USA',), True),
        Testcase(('leetcode',), True),
        Testcase(('Google',), True),
        Testcase(('gG',), False),
        Testcase(('FlaG',), False),
        Testcase(('flaG',), False),
    ))

File: TypeScript/solutions/detect-capital.ts
// https://leetcode.com/problems/detect-capital/description/
import { Examiner, Testcase } from '../utils';


function detectCapitalUse(word: string): boolean {
    if (word.length < 2) return true;
    if (word[0]! === word[0]!.toLowerCase() && word[1]! !== word[1]!.toLowerCase()) return false;
    const isLower = word[1]! === word[1]!.toLowerCase();
    for (let i = 2; i < word.length; i++) {
        if ((word[i]! === word[i]!.toLowerCase()) !== isLower) return false;
    }
    return true;
}

export function examine(): void {
    new Examiner(detectCapitalUse).run([
        new Testcase([''], true),
        new Testcase(['g'], true),
        new Testcase(['gg'], true),
        new Testcase(['G'], true),
        new Testcase(['USA'], true),
        new Testcase(['leetcode'], true),
        new Testcase(['Google'], true),
        new Testcase(['gG'], false),
        new Testcase(['FlaG'], false),
        new Testcase(['flaG'], false),
    ]);
}
